### PR TITLE
Deterministic pagination in random albums sort

### LINKED ISSUE #749: Random sort changes each time you switch pages
In random sort mode, clicking the next page button changes the random ordering, so some of the same albums show up. Maybe a random seed should be stored in the url?

### PR DESCRIPTION
**Pull request by gqgs**
Register a new `RAND` SQLite function and use it when paginating random albums to avoid duplicate results.
This new function returns a pseudo-random value based on a seed and the album's unique identifier.

Fixes #749

### CODE REVIEW

File: utils/strings.go [hunk 1]
```diff
@@ -79,3 +80,14 @@ func LongestCommonPrefix(list []string) string {
...
+func StringHasher() func(str string) uint64 {
+	seed := maphash.MakeSeed()
+	var hash maphash.Hash
+	hash.SetSeed(seed)
+	return func(str string) uint64 {
+		hash.Reset()
+		_, _ = hash.WriteString(str)
+		return hash.Sum64()
+	}
+}
```
Comment (gqgs) [inline utils/strings.go:93]: Could be more efficient and simpler if we were using Go 1.19.
```go
func StringHasher() func(str string) uint64 {
	seed := maphash.MakeSeed()
	return func(str string) uint64 {
		return maphash.String(seed, str)
	}
}
```
https://pkg.go.dev/hash/maphash@master#String

File: utils/strings.go [hunk 2]
```diff
@@ -79,3 +80,14 @@ func LongestCommonPrefix(list []string) string {
 	}
 	return list[0]
 }
+
+func StringHasher() func(str string) uint64 {
+	seed := maphash.MakeSeed()
```
Comment (gqgs) [inline utils/strings.go:85]: We may need a way to re-instantiate a new seed.

File: utils/strings.go (hunk 1 above)
Comment (deluan) [reply utils/strings.go:93]: We should keep compatibility with Go 1.18 until at least 1.20 is out. We have to be able to build binaries for platforms that upgrades in a slower pace.

File: utils/strings.go (hunk 2 above)
Comment (deluan) [reply utils/strings.go:85]: I loved your solution, very simple! But without fixing this re-seed, we cannot use it, or else when we refresh the Random Albums page we get the same result.... We have to figure this out.

Comment (gqgs) [reply utils/strings.go]: Yeah I agree. Just not sure what approach we want to take here to get us closer to the ideal behavior.
A pushed a new commit to reseed at the "first page" of the request for random albums.
It should be enough to fix the issue you mentioned and the one linked in the PR description.

**Comment by deluan** [comment]
Ok, I testd it with a huge library (~1M songs), it works and it is fast! Great idea and great job! But I see some issues:
1) That is easy, we need to use the new rand() funcion for the Subsonic API's `getAlbumList` as well. This is simple and it is a matter of changing this line https://github.com/navidrome/navidrome/blob/5621551dd09eb4a5807a509cca98f0a77493f331/server/subsonic/filter/filters.go#L25 to
`return Options{Sort: "random"}` (without the `()`)
2) We may use this in other places, like random songs and the `random` sorting in Smart Playlists, but this can be a separated PR.
3) It would be great to have some tests ;)

But the real problem is: Navidrome is multiuser and the `hasher.Reseed()` method could be called by another user while we are paginating, this would cause our own pagination to be reseeded as well. We need to find a way to have one hasher per user. Maybe put it in the context? In this case, how we pass the context value to the function called by SQLite? That could make things more complicated than it should be.... Any ideas?

**Comment by gqgs** [comment]
@deluan Thanks for the feedback. I pushed some commits to use the function in the Subsonic API and to implement some unit tests.

>But the real problem is: Navidrome is multiuser

Good call. I didn't think about that.
I'm not seeing a great solution either for that case. I'll post an update here if I think of something.

**Comment by deluan** [comment]
Hey @gqgs, after a looooong time, I gave this PR another close look, and found a way to have a seed per user. If you can, please check my last commit (f59cb50) and let me know what you think.

**Pull request merged**